metrics_utility/automation_controller_billing/collectors.py

```diff
@@ -302,3 +299,52 @@ def main_jobevent_table(since, full_path, until, **kwargs):
     return _copy_table(table=tbl,
                        query=f"COPY ({query}) TO STDOUT WITH CSV HEADER",
                        path=full_path)
+
+@register('indirect_nodes', '1.0', format='csv', description=_('Data for billing'), fnc_slicing=daily_slicing)
+def indirect_nodes_table(since, full_path, until, **kwargs):
+
+    if not INCLUDE_INDIRECT:
+        return None
+
+    query = f"""
+        (
+            SELECT
+                main_indirectmanagednodeaudit.id,
+                main_indirectmanagednodeaudit.created as created,
+                main_indirectmanagednodeaudit.name as host_name,
+                main_indirectmanagednodeaudit.host_id AS host_remote_id,
+                main_indirectmanagednodeaudit.canonical_facts,
+                main_indirectmanagednodeaudit.facts,
+                main_indirectmanagednodeaudit.events,
+                main_indirectmanagednodeaudit.count as task_runs,
+                main_unifiedjob.created AS job_created,
+                main_indirectmanagednodeaudit.job_id AS job_remote_id,
+                main_unifiedjob.unified_job_template_id AS job_template_remote_id,
+                main_unifiedjob.name AS job_template_name,
+                main_inventory.id AS inventory_remote_id,
+                main_inventory.name AS inventory_name,
+                main_organization.id AS organization_remote_id,
+                main_organization.name AS organization_name,
+                main_unifiedjobtemplate_project.id AS project_remote_id,
+                main_unifiedjobtemplate_project.name AS project_name
+            FROM main_indirectmanagednodeaudit
+            LEFT JOIN main_job
+                ON main_job.unifiedjob_ptr_id = main_indirectmanagednodeaudit.job_id
+            LEFT JOIN main_unifiedjob
+                ON main_unifiedjob.id = main_indirectmanagednodeaudit.job_id
+            LEFT JOIN main_inventory
+                ON main_inventory.id = main_indirectmanagednodeaudit.inventory_id
+            LEFT JOIN main_organization
+                ON main_organization.id = main_unifiedjob.organization_id
+            LEFT JOIN main_unifiedjobtemplate AS main_unifiedjobtemplate_project
+                ON main_unifiedjobtemplate_project.id = main_job.project_id
+            WHERE (main_indirectmanagednodeaudit.created >= '{since.isoformat()}'
+            AND  main_indirectmanagednodeaudit.created < '{until.isoformat()}')
```

Comment:
don't we have some decent way of escaping things?

This is safe as long as they're datetime objects, but otherwise that's how you get sql injections.

Reply:
@himdel  yes, but it's like that in all collectors I think. (including the one in controller) This has only internal access, where you can already access DB creds and have full access then.

That being said, it safer to not have the code sql injectable (although the isoformat will break sql injection attempt). We should just use psycopg, like we do it e.g. in analytics https://gitlab.cee.redhat.com/automation-analytics/automation-analytics-backend/-/blob/main/restapi/v1/content_explorer/handlers_content_explorer.py#L47

See https://www.psycopg.org/psycopg3/docs/api/sql.html#module-usage, that offers usage for Literal and indentifier, for correct escaping of queries.